### PR TITLE
math: trig: Improve trig function execution speed

### PR DESCRIPTION
**Pull request by ShriramShastry**
if conditions in loops slow down the execution speed.So
 to optimize the loop containing if conditions where branch
 is separated by sign only. A new arithmetic math added .

 Added new defines to replace Q_CONVERT_FLOAT constants.

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>

### CODE REVIEW

**Review by paulstelian97 — APPROVED** [review]

**Review by singalsu — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kv2019i @lyakh fyi - looks like this causes a section overlap in Zephyr linking. Do we need to update the toml ?
```
+ RIMAGE_DIR=build-rimage
  Found 45 sections, listing valid sections......
+ cmake -B build-rimage -S modules/audio/sof/rimage
	No	Start		End		Size	Type	Name
	1	0xbe010000	0xbe01016a	0x16a	TEXT	.WindowVectors.text
	2	0xbe010180	0xbe0101a9	0x29	TEXT	.Level2InterruptVector.text
	3	0xbe0101c0	0xbe0101e9	0x29	TEXT	.Level3InterruptVector.text
	4	0xbe010200	0xbe010229	0x29	TEXT	.Level4InterruptVector.text
	5	0xbe010240	0xbe010269	0x29	TEXT	.Level5InterruptVector.text
	6	0xbe010280	0xbe0102a9	0x29	TEXT	.DebugExceptionVector.text
	7	0xbe0102c0	0xbe0102e9	0x29	TEXT	.NMIExceptionVector.text
	8	0xbe010300	0xbe010303	0x3	TEXT	.KernelExceptionVector.text
	9	0xbe010340	0xbe010352	0x12	TEXT	.UserExceptionVector.text
	10	0xbe0103c0	0xbe0103c6	0x6	TEXT	.DoubleExceptionVector.text
	11	0xbe010400	0xbe031000	0x20c00	TEXT	.text
	12	0xbe031000	0xbe0452b0	0x142b0	DATA	.rodata
	13	0xbe0452b0	0xbe045318	0x68	DATA	initlevel
	14	0xbe045318	0xbe0453a8	0x90	DATA	devices
	15	0xbe0453a8	0xbe0453c0	0x18	DATA	k_p4wq_initparam_area
	16	0xbe0453c0	0xbe0453f8	0x38	DATA	log_const_sections
	17	0xbe0453f8	0xbe045408	0x10	DATA	log_backends_sections
	18	0xbe045408	0xbe045444	0x3c	DATA	device_handles
	19	0xbe045444	0xbe0454b0	0x6c	DATA	.fw_ready
	20	0x9e0454c0	0x9e045960	0x4a0	DATA	.noinit
	21	0x9e045980	0x9e047000	0x1680	DATA	.data
	22	0xbe0454c0	0xbe04fcc0	0xa800	DATA	.cached
	23	0x9e047000	0x9e0474a8	0x4a8	DATA	sw_isr_table
	24	0x9e0474a8	0x9e0474c0	0x18	DATA	device_states
	25	0x9e0474c0	0x9e0474d0	0x10	DATA	k_sem_area
	26	0x9e048000	0x9e1fae50	0x1b2e50	BSS	.bss
	39	0x1fffa000	0x1fffa630	0x630	NOTE	.static_uuid_entries
	40	0x20000000	0x20000060	0x60	NOTE	.static_log_entries
	41	0x22000000	0x220002d0	0x2d0	NOTE	.fw_metadata

 module: input size 269083 (0x41b1b) bytes 26 sections
 module: text 134779 (0x20e7b) bytes
    data 134304 (0x20ca0) bytes
    bss  1781328 (0x1b2e50) bytes

fw abi main version:	3.19.0
fw abi dbg version:	5.3.0
logs dictionary size:	172
including fw version of size:	60
uids dictionary size:	1600
-- Configuring done
-- Generating done
-- Build files have been written to: /workdir/zephyrproject/build-rimage
+ make -C build-rimage -j2
make: Entering directory '/workdir/zephyrproject/build-rimage'
make[1]: Entering directory '/workdir/zephyrproject/build-rimage'
make[2]: Entering directory '/workdir/zephyrproject/build-rimage'
Consolidate compiler generated dependencies of target rimage
make[2]: Leaving directory '/workdir/zephyrproject/build-rimage'
[100%] Built target rimage
make[1]: Leaving directory '/workdir/zephyrproject/build-rimage'
make: Leaving directory '/workdir/zephyrproject/build-rimage'
+ west sign --build-dir build-icl --tool rimage --tool-path build-rimage/rimage --tool-data modules/audio/sof/rimage/config -- -k modules/audio/sof/keys/otc_private_key.pem
Signing for board intel_adsp_cavs20
Signing for SOC target icl using icl.toml
Signing with tool build-rimage/rimage
build-rimage/rimage -k modules/audio/sof/keys/otc_private_key.pem -o /workdir/zephyrproject/build-icl/zephyr/zephyr.ri -c modules/audio/sof/rimage/config/icl.toml -i 3 -e /workdir/zephyrproject/build-icl/zephyr/bootloader.elf.mod /workdir/zephyrproject/build-icl/zephyr/zephyr.elf.mod
error: section overlap between /workdir/zephyrproject/build-icl/zephyr/zephyr.elf.mod:20 and /workdir/zephyrproject/build-icl/zephyr/zephyr.elf.mod:22
     [0xbe0454c0 : 0xbe045960] overlaps with [0xbe0454c0 :0xbe04fcc0]

Module Reading /workdir/zephyrproject/build-icl/zephyr/bootloader.elf.mod
info: ignore .bss section for bootloader module
  Found 21 sections, listing valid sections......
	No	LMA		VMA		End		Size	Type	Name
	0	0x00000000	0x00000000	0x00000000	0x0
	1	0xb0038000	0xb0038000	0xb00380fb	0xfb	TEXT	.boot_entry.text
FATAL ERROR: command exited with status 234: build-rimage/rimage -k modules/audio/sof/keys/otc_private_key.pem -o /workdir/zephyrproject/build-icl/zephyr/zephyr.ri -c modules/audio/sof/rimage/config/icl.toml -i 3 -e /workdir/zephyrproject/build-icl/zephyr/bootloader.elf.mod /workdir/zephyrproject/build-icl/zephyr/zephyr.elf.mod
	2	0xb0038220	0xb0038220	0xb00384c2	0x2a2	TEXT	.text
	3	0xb0039000	0xb0039000	0xb0039008	0x8	DATA	.rodata

 module: input size 933 (0x3a5) bytes 4 sections
 module: text 925 (0x39d) bytes
    data 8 (0x8) bytes
    bss  0 (0x0) bytes


Module Reading /workdir/zephyrproject/build-icl/zephyr/zephyr.elf.mod
  Found 46 sections, listing valid sections......
	No	LMA		VMA		End		Size	Type	Name
	1	0xbe010000	0xbe010000	0xbe01016a	0x16a	TEXT	.WindowVectors.text
	2	0xbe010180	0xbe010180	0xbe0101a9	0x29	TEXT	.Level2InterruptVector.text
	3	0xbe0101c0	0xbe0101c0	0xbe0101e9	0x29	TEXT	.Level3InterruptVector.text
	4	0xbe010200	0xbe010200	0xbe010229	0x29	TEXT	.Level4InterruptVector.text
	5	0xbe010240	0xbe010240	0xbe010269	0x29	TEXT	.Level5InterruptVector.text
	6	0xbe010280	0xbe010280	0xbe0102a9	0x29	TEXT	.DebugExceptionVector.text
	7	0xbe0102c0	0xbe0102c0	0xbe0102e9	0x29	TEXT	.NMIExceptionVector.text
	8	0xbe010300	0xbe010300	0xbe010303	0x3	TEXT	.KernelExceptionVector.text
	9	0xbe010340	0xbe010340	0xbe010352	0x12	TEXT	.UserExceptionVector.text
	10	0xbe0103c0	0xbe0103c0	0xbe0103c6	0x6	TEXT	.DoubleExceptionVector.text
	11	0xbe010400	0xbe010400	0xbe031000	0x20c00	TEXT	.text
	12	0xbe031000	0xbe031000	0xbe0452b0	0x142b0	DATA	.rodata
	13	0xbe0452b0	0xbe0452b0	0xbe045318	0x68	DATA	initlevel
	14	0xbe045318	0xbe045318	0xbe0453a8	0x90	DATA	devices
	15	0xbe0453a8	0xbe0453a8	0xbe0453c0	0x18	DATA	k_p4wq_initparam_area
	16	0xbe0453c0	0xbe0453c0	0xbe0453f8	0x38	DATA	log_const_sections
	17	0xbe0453f8	0xbe0453f8	0xbe045408	0x10	DATA	log_backends_sections
	18	0xbe045408	0xbe045408	0xbe045444	0x3c	DATA	device_handles
	19	0xbe045444	0xbe045444	0xbe0454b0	0x6c	DATA	.fw_ready
	20	0xbe0454c0	0xbe0454c0	0xbe045960	0x4a0	DATA	.noinit
	21	0xbe045980	0xbe045980	0xbe047000	0x1680	DATA	.data
	22	0xbe0454c0	0xbe0454c0	0xbe04fcc0	0xa800	DATA	.cached
	23	0xbe047000	0xbe047000	0xbe0474a8	0x4a8	DATA	sw_isr_table
	24	0xbe0474a8	0xbe0474a8	0xbe0474c0	0x18	DATA	device_states
	25	0xbe0474c0	0xbe0474c0	0xbe0474d0	0x10	DATA	k_sem_area
	26	0xbe048000	0xbe048000	0xbe1fae50	0x1b2e50	BSS	.bss

 module: input size 269083 (0x41b1b) bytes 26 sections
 module: text 134779 (0x20e7b) bytes
    data 134304 (0x20ca0) bytes
    bss  1781328 (0x1b2e50) bytes
```

**Comment by lgirdwood** [comment]
@lyakh any comment here, the change should only be adding 10s of bytes AFAICT.

File: src/math/trig.c [hunk 1]
```diff
@@ -210,10 +200,10 @@ int32_t is_scalar_cordic_asin(int32_t sinvalue, int16_t numiters)
...
 		x = 0;
-		y = cord_arcsincos_q30fl;
+		y = CORD_ARCSINCOS_Q2_30;
 		z = PI_DIV2_Q3_29;
 	} else {
-		x = cord_arcsincos_q30fl;
+		x = CORD_ARCSINCOS_Q2_30;
 		y = 0;
 		z = 0;
 	}
```
Comment (lyakh) [inline src/math/trig.c:209]: just noticed some code below:
```
	if (j >= 31)
		j = 31;
```
hopefully the compiler catches this, but you obviously don't need to assign 31, if `j` is already 31, so the condition should be `>`, not `>=` Also, if it's important to optimise every bit of it, you could use the fact, that if `b_i > 31` then `j` is definitely `> 31` too. You could also use `__buiktin_expect()` (often wrapped in `likely()` / `unlikely()` style macros) to help the compiler optimise for more likely branches.

**Comment by lyakh** [comment]
> @lyakh any comment here, the change should only be adding 10s of bytes AFAICT.

interesting... Looks like the linker script is wrong - `.cached` and `.noinit` have the same start address... I don't think this is caused by this PR

**Comment by lgirdwood** [comment]
> > @lyakh any comment here, the change should only be adding 10s of bytes AFAICT.
> 
> interesting... Looks like the linker script is wrong - `.cached` and `.noinit` have the same start address... I don't think this is caused by this PR

@andyross @dcpleung fyi looks like a linker script issue if cached and noinit have the same address ?
@lyakh can you take a quick looks here for anything obvious.

**Comment by dcpleung** [comment]
Should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/40319.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Looks like an unrelated failure, so rerun CI.

**Pull request closed without merging**